Development

ajatdarojat45 opened the pull request:
This pull request delivers a significant update focused on enhancing the user interface and user experience of the application, particularly around project and model management, sidebar navigation, and overall visual styling. Key improvements include a major redesign of the `ProjectCard` and `ModelCard` components, a new `WelcomeSidebar` for onboarding, and the addition of a sources & receivers sidebar tab. Several bug fixes and minor UI tweaks are also included.

**UI/UX Enhancements:**

- Redesigned `ProjectCard` and `ModelCard` components with improved layouts, gradient borders, model image displays, and dropdown menus for actions like deletion. This provides a more modern and visually appealing interface. [[1]](diffhunk://#diff-d329be54503541c0e8039d7457d6455d0ad697995adb990ce87489ab71a633f2L50-R61) [[2]](diffhunk://#diff-d329be54503541c0e8039d7457d6455d0ad697995adb990ce87489ab71a633f2L95-R119) [[3]](diffhunk://#diff-94ad9a358cff117dd638812e32d57b91b98ab7e3a76e8800cae643b67b955f59L2-R17) [[4]](diffhunk://#diff-94ad9a358cff117dd638812e32d57b91b98ab7e3a76e8800cae643b67b955f59L26-R83)
- Added a new `WelcomeSidebar` component featuring a GitHub icon and quick links to the repository and documentation, improving onboarding and navigation for new users.
- Updated `EmptySimulation`, `GroupPicker`, and form components with enhanced colors, text styles, and button consistency for a more cohesive look. [[1]](diffhunk://#diff-37f3699ae269c0d145eab6e2ba6ce25a6dbfcf1cfe33d3f792af4e83b871d2fbL12-R14) [[2]](diffhunk://#diff-6da393fb54f43b5abf883849bdb5531890531896d8cafaaa7e57beaa26e8940cL19-R23) [[3]](diffhunk://#diff-4b38bb0604cb28928df515be063c4be0d5b5f65046b88b50b4c54d5a1688c7ebL69-R69) [[4]](diffhunk://#diff-4b38bb0604cb28928df515be063c4be0d5b5f65046b88b50b4c54d5a1688c7ebL127-R127) [[5]](diffhunk://#diff-4b38bb0604cb28928df515be063c4be0d5b5f65046b88b50b4c54d5a1688c7ebL221-R220) [[6]](diffhunk://#diff-18bf4ab6ed5d5f1afa9bff3a23b7b680213bc5b0d2f5f3944be38c7ceda99101R47-R50) [[7]](diffhunk://#diff-18bf4ab6ed5d5f1afa9bff3a23b7b680213bc5b0d2f5f3944be38c7ceda99101L133-R134)

**Sidebar and Navigation Improvements:**

- Refactored and renamed the sidebar tabs to `simulationSettings/SidebarTabs.tsx`, introduced a new `SourceReceiversTab`, and adjusted tab layout for better responsiveness and clarity. [[1]](diffhunk://#diff-5128dcd3f1e952da8b3ad4b29445dbf3b533c60fee8482f4ef33c96c751563ffR3-R15) [[2]](diffhunk://#diff-5128dcd3f1e952da8b3ad4b29445dbf3b533c60fee8482f4ef33c96c751563ffL30-R26)
- Added a new `SourceReceiversMenu` dropdown for managing sources and receivers directly from the sidebar.

**General and Infrastructure:**

- Updated to version 0.0.10 in `package.json` and added a detailed changelog entry summarizing all features and fixes in this release. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L4-R4) [[2]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR5-R34)
- Improved font loading and layout responsiveness by preconnecting to Google Fonts in `index.html`.

**Bug Fixes:**

- Fixed validation and layout issues in various components, including receiver updates and header responsiveness.

**Summary of Most Important Changes:**

**1. UI/UX Redesign and Enhancements**
- Major redesign of `ProjectCard` and `ModelCard` components, including gradient borders, image previews, and dropdown action menus for improved usability and aesthetics. [[1]](diffhunk://#diff-d329be54503541c0e8039d7457d6455d0ad697995adb990ce87489ab71a633f2L50-R61) [[2]](diffhunk://#diff-d329be54503541c0e8039d7457d6455d0ad697995adb990ce87489ab71a633f2L95-R119) [[3]](diffhunk://#diff-94ad9a358cff117dd638812e32d57b91b98ab7e3a76e8800cae643b67b955f59L2-R17) [[4]](diffhunk://#diff-94ad9a358cff117dd638812e32d57b91b98ab7e3a76e8800cae643b67b955f59L26-R83)
- Enhanced form and button styling across the app for consistency and clarity. [[1]](diffhunk://#diff-4b38bb0604cb28928df515be063c4be0d5b5f65046b88b50b4c54d5a1688c7ebL69-R69) [[2]](diffhunk://#diff-4b38bb0604cb28928df515be063c4be0d5b5f65046b88b50b4c54d5a1688c7ebL127-R127) [[3]](diffhunk://#diff-4b38bb0604cb28928df515be063c4be0d5b5f65046b88b50b4c54d5a1688c7ebL221-R220) [[4]](diffhunk://#diff-18bf4ab6ed5d5f1afa9bff3a23b7b680213bc5b0d2f5f3944be38c7ceda99101R47-R50) [[5]](diffhunk://#diff-18bf4ab6ed5d5f1afa9bff3a23b7b680213bc5b0d2f5f3944be38c7ceda99101L133-R134)

**2. Sidebar and Navigation Improvements**
- Refactored sidebar tabs, introduced the `SourceReceiversTab` for managing sources and receivers, and improved sidebar layout and responsiveness. [[1]](diffhunk://#diff-5128dcd3f1e952da8b3ad4b29445dbf3b533c60fee8482f4ef33c96c751563ffR3-R15) [[2]](diffhunk://#diff-5128dcd3f1e952da8b3ad4b29445dbf3b533c60fee8482f4ef33c96c751563ffL30-R26)
- Added `SourceReceiversMenu` for quick actions on sources/receivers.

**3. Onboarding and Documentation Access**
- Added a new `WelcomeSidebar` with GitHub and documentation links to help users get started and access resources easily.

**4. General Infrastructure Updates**
- Updated app version to 0.0.10 and added a comprehensive changelog entry. [[1]](diffhunk://#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L4-R4) [[2]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR5-R34)
- Improved font loading performance and layout responsiveness in `index.html`.

**5. Bug Fixes and Minor Tweaks**
- Fixed validation and layout bugs in group selection, receivers, and header components for a smoother user experience. [[1]](diffhunk://#diff-06572a96a58dc510037d5efa622f9bec8519bc1beab13c9f251e97e657a9d4edR5-R34) [[2]](diffhunk://#diff-6da393fb54f43b5abf883849bdb5531890531896d8cafaaa7e57beaa26e8940cL19-R23) [[3]](diffhunk://#diff-37f3699ae269c0d145eab6e2ba6ce25a6dbfcf1cfe33d3f792af4e83b871d2fbL12-R14)

Overall, this release brings a more polished, user-friendly, and visually consistent interface, while laying the groundwork for improved simulation and project management workflows.

Commit:
chore: tidying up components into some folders

Commit:
feat: material list with simple search

ajatdarojat45 committed:
Merge pull request #22 from tigadika/feature/materialEditor

Feature/material editor

Commit:
feat: add sources & receivers functionality in side tab

Commit:
feat: select source point feature

Commit:
fix: transform control rendered double on group. feat: add render receivers functionality

Commit:
feat: wiring to api for source and receiver

fachrihawari committed:
feat: enhance app layout and styling with sidebar adjustments and font integration

fachrihawari committed:
feat: update app layout header styling and add custom color variables

fachrihawari committed:
fix: correct font class for header title in AppLayout component

fachrihawari committed:
fix: adjust header layout and improve link structure in AppLayout component

fachrihawari committed:
feat: enhance ProjectForm and AppLayout components with improved button styles and labels

fachrihawari committed:
feat: add WelcomeSidebar component and integrate GitHub icon; update button styles in ProjectForm and UploadModel

fachrihawari committed:
fix: adjust layout properties in AppLayout component for better responsiveness

fachrihawari committed:
feat: enhance project management features with group selection and improved UI components

fachrihawari committed:
feat: add plus icon to ProjectForm trigger in HomePage component

fachrihawari committed:
feat: update ModelCard and UploadModel components for improved UI and functionality; add model image display and delete confirmation

fachrihawari committed:
feat: set default values for ProjectForm in HomePage and update group fallback to 'NONE'

fachrihawari committed:
feat: refactor ModelCard component to use Card and DropdownMenu for improved UI and delete functionality

fachrihawari committed:
feat: add upload icon to ProjectDetailPage for improved user experience

fachrihawari committed:
feat: update EditorPage title to display model project and name with navigation link

fachrihawari committed:
feat: enhance EmptySimulation component styling and update SidebarTabs for responsive layout

fachrihawari committed:
feat: update ProjectCard to display up to three model images with adjusted positioning and styling

Commit:
fix: add validation to update receivers

Commit:
feat: add settings tab, fetch settings, and update simulation

Commit:
feat: add monaco editor to edit setting with json

ajatdarojat45 committed:
Merge pull request #23 from tigadika/feature/source-receiver

Feature/source receiver

ajatdarojat45 committed:
Merge pull request #24 from fachrihawari/feat/layout-home-detail-project

feat: Layout Home & Detail project

ajatdarojat45 committed:
chore(release): 0.0.10

ajatdarojat45 commented:
Hi @SilvinWillemsen,

This PR includes several new updates and improvements:
- Added sources to the 3D viewport and database.
- Added receivers to the 3D viewport and database.
- Implemented transform controls for sources and receivers, allowing objects to be moved directly in the viewport.
- Updated the Home and Project Detail layouts based on the latest Figma design.
- Update project and change the group of project.

Please have a look and let me know your review and feedback when you get a chance.

cc: @fachrihawari @tigadika

SilvinWillemsen commented:
Dear @ajatdarojat45 (& @fachrihawari, as I see you worked a lot on this PR :)),

Thanks for this great update! It's starting to really look good and feel very usable. I'll hold off of giving feedback on the visuals / styling for now, as I think this has lower priority. Below you can find my feedback for the functionality:

Projects menu
- A group seems to be automatically deleted if there are no projects assigned to it. Is this intentional? 
    - (Low priority:) Perhaps we could still show groups that don’t have projecs, but have a three-dot button (or a trash button) next to the group name to allow for deletion of groups. Possibly we could have a popup when clicking this trash button, asking whether you want to delete all projects assigned to that group. Then if “no” is selected, those projects automatically get assigned to the “No group” group.

Model viewport & Sources/receivers
- Clicking anywhere in the viewport should deselect all sources and receivers.
- Selecting a source changes its colour from blue to orange. Let's stick with blue for now :)
- The way you implemented the errors are nice! A few comments:
    - The errors only seem to be implemented for a receiver. You can add the same errors to the source as well.
    - The surface error seems to be more important than the “outside the model” error, which is not what I would expect as a user. (try to place a receiver right outside the model close to a surface, I get the surface error, not the outside the model error) Errors in order of importance: “outside model” → “too close to surface” → source/receiver distance.
    - Some of these errors (e.g. the surface error) will be simulation-method dependent. While coding, you could maybe keep this in mind.
- Clicking on the textboxes (or the arrows in them) currently (de)selects the source/receiver. Could this be disabled?
- The “Select Source/Receiver here to edit position” text is unnecessary
- The sources start at 0 and the receivers at 1. Both should start at 1.
- Could we have a row below the sources / receivers with a “+”? This way the user has easy access to adding sources and receivers. I’ll make a design for this soon. Perhaps you could temporarily increase the maximum sources/recievers to 5 for testing.

Thanks again and let me know if you have any questions :)

ajatdarojat45 commented:
Hi @SilvinWillemsen 

Yes, that behavior (A group seems to be automatically deleted) is intentional — the group data is aggregated directly from the existing project data, and there’s currently no group data coming from the server.

If we’d like to persist groups even when no projects are assigned, there are two possible approaches:
	1.	Local state — but the data would be lost upon page refresh.
	2.	Local storage — persists on the client side (browser).

For consistency, I’d recommend going with the second option (local storage).

Cc: @fachrihawari

Commit:
fix: number order for sources not start at 1

Commit:
fix: select outside the sources/receivers, reset selection

Commit:
fix: select receiver deselect source and vice versa

Commit:
fix: conflict validation outside model and close to surface

Commit:
fix: add handle propagation in all input source receivers

Commit:
fix: add validation to source too, not just receivers

SilvinWillemsen commented:
Ah alright, makes sense. Then I think it would be best to create a separate entry for Group in the database actually. What do you think? 

For now, this has low priority, though.

fachrihawari committed:
feat: implement DeleteGroup component with project deletion functionality

ajatdarojat45 commented:
Hi @SilvinWillemsen,

Got it!
Yes, I agree that having a dedicated Group entry in the database would be ideal in the long run.
For now, since it’s low priority, we’ll keep persisting the group data only on the client side (in the browser using local storage).

fachrihawari committed:
fix: enhance "No projects found" message layout for better user experience

Commit:
fix: move add button to bottom row

fachrihawari committed:
feat: add group management functions for project deletion and storage

Commit:
style: match style in sources receivers tab to matche design guidelines

fachrihawari committed:
feat: integrate Redux for group management and remove localStorage dependency

fachrihawari committed:
fix: improve sorting logic for unique groups in project selector

fachrihawari committed:
feat: add projectsCount prop to DeleteGroup component for improved project management

Commit:
style: change some styling in editor page

ajatdarojat45 committed:
Merge pull request #26 from fachrihawari/feat/delete-group

Feat/delete group

ajatdarojat45 committed:
Merge pull request #27 from tigadika/fix/sources-receivers

Fix/sources receivers

ajatdarojat45 committed:
chore(release): 0.0.11

ajatdarojat45 commented:
Hi @SilvinWillemsen,

Thanks a lot for the detailed feedback! 
All mentioned points have been addressed and pushed:
- Clicking anywhere now deselects sources/receivers.
- Selecting a source keeps the blue color.
- Added error validation for sources as well.
- Updated validation logic for “outside model” and “surface too close” calculations.
- Added stopPropagation in input fields to prevent (de)selection.
- Removed the “Select Source/Receiver…” placeholder text.
- Sources now start from 1 by default.
- Added a “+” button at the bottom of the row to add new sources/receivers.
- Updated styling to match the Editor page design.
- Added Delete group functionality.

Please have a look and let we know your review or if there’s anything else we should refine.

SilvinWillemsen commented:
Dear @ajatdarojat45 & @fachrihawari,

Thanks for a great update! Some comments:

**Sources/receivers**
- The source doesn't seem to have the "Too close to a receiver" error. This is alright, but moving the source should at least update the receiver. 
    - Currently: 
        - Source and receiver are far apart.
        - Source gets moved close to receiver.
        - Receiver doesn't show error
    - Desired behaviour: receiver shows error in step 3.
- The sources and receivers should be points, i.e., a 0-dimensional location at the centre of the visualised sphere. However, even when a tiny part of the sphere is outside the model bounds, the "outside the model bounds" error gets shown. Is it possible to only check the "outside the model" against the centre of the sphere, not the volume (as currently happens)?
- Nice that you started implementing the proper tabs design! Could you centre the text horizontally such that the text in the non-active buttons is centred between the side panel and the right side of the button?

**Groups**
The group handling seems to work nice now! Few comments here:
- The trashcan doesn't have to be so obvious (perhaps black with 50% opacity, and 75% on hover?) Also the underline only has to go underneath the text, not the trashcan, and can also be 50% opacity.
- When I want to create a new group and start typing the name of an existing group, I can't seem to type it (see [video](https://github.com/user-attachments/assets/7ba37757-4e64-4703-9862-8de9a9f8d6c9)). Also my password manager seems to recognise many input boxes as "username" or "password" boxes. I'm not sure how this works, but do you think this can be disabled?

**Other comments**
- After creating a new simulation, this is not automatically reflected in the SimulationPicker dropdown. Could this be implemented?
- The Geometry information doesn't need a gradient border; this can just be grey as before.
- On the topic of borders, maybe the borders of all the cards (project/model) can be grey until hovered. Also the + in the + cards should be grey until hovered. (Low priority, but if it's an easy thing to implement you can do it :))

Thanks for the great work! Let me know when you have a new update I can look at :)

fachrihawari committed:
style: improve button styling and layout in DeleteGroup component

fachrihawari committed:
style: enhance CreateGroup component by preventing dialog close on keydown and improving input accessibility

fachrihawari committed:
feat: add lazy query for simulations by model ID and update cache invalidation tags

ajatdarojat45 commented:
Hi @SilvinWillemsen,

Thanks a lot for the detailed feedback!
Regarding the smaller visual and styling details (like colors, borders, alignment, and icon opacity) — would it be okay if we address those a bit later, once the main functionality is fully stable?

This way, we can keep the review process smoother and focus on the functional parts first, then handle the design refinements together in one go toward the end.

SilvinWillemsen commented:
Hi @ajatdarojat45, yes all good! I just wanted to mention it in case it was easier to do it right away. Do whatever seems best for you!

tigadika commented:
Hi @SilvinWillemsen , 

sorry to jump in. I'm one of the developer responsible for editor page & the 3D part, including the source and the receiver. The reason you don't see my name in the commit history because I forgot to setup my github profile in the project before :smile: . I just want to confirm the validation in the sources & receivers:

- for the outside-model validation calculation, it only counts when the center point crosses the axis bounds, right?
- If that’s the case, would it be okay if the “too close to the surface” validation is triggered only when the location point equals the model’s axis bounds? That way, the calculation could be simplified, because the current implementation considers both the point volume and a threshold margin (±), which makes it more complex.

thankyou!
<img width="1280" height="782" alt="image" src="https://github.com/user-attachments/assets/c299dd75-ae22-4980-94ad-560c22a51e21" />

SilvinWillemsen commented:
Hi @tigadika,

Right! Thanks for your explanation and sorry for not tagging you before :) Great work already on the 3D viewport + editor! It’s really looking good 💪🏼

Just for some background: the “too close to surface” error is needed for some simulation methods, as they won’t work if the source / receiver is too close to a surface. The radius of the validation sphere also depends on the settings of this simulation method. I’ll double check, but I think the DG method (and possibly others) need this validation check. 

Rather than using the visible sphere for validation, imagine an invisible sphere of (say for now) 0.2m around the source / receiver. This should be validated, not the visible sphere.

Thanks for your drawing also, that makes things a lot clearer :) I propose the following:
x <= 0: outside the model error
0 < x < 0.2: too close to surface error
x >= 0.2: no error

Of course the validation is in 3D, but I’m sure that’s clear :)

Let me know if this makes sense!

tigadika committed:
fix: change validation logic

tigadika committed:
fix: change validation logic

tigadika committed:
fix: change validation logic

tigadika committed:
fix: validate every source and receivers proximity

tigadika committed:
style: change styling tab menu

tigadika committed:
style: change styling geometry info

ajatdarojat45 committed:
Merge pull request #29 from fachrihawari/fix/group-trash

fix: Group trash and minor bug

ajatdarojat45 committed:
Merge pull request #30 from tigadika/fix/sources-receivers

Fix/sources receivers

ajatdarojat45 committed:
chore(release): 0.0.12

ajatdarojat45 commented:
Hi @SilvinWillemsen,

We’ve pushed several updates based on your comments:
- Style: Improved button styling and layout in the DeleteGroup component.
- Style: Enhanced CreateGroup component — the dialog now doesn’t close on keydown, and input accessibility has been improved.
- Feat: Added lazy query for simulations by model ID and updated cache invalidation tags.
- Fix: Updated validation logic for sources and receivers.
- Fix: Each source and receiver now validates proximity correctly.
- Style: Adjusted styling for the tab menu and geometry information section.

Please have a look and let us know if everything works as expected or if there’s anything else you’d like us to refine. 

cc: @fachrihawari @tigadika

tigadika committed:
Merge branch 'development' into feature/setting-tab

tigadika committed:
fix: change simulation method refetch the settings and populate docs button

tigadika committed:
fix: change simulation actually change the settings

tigadika committed:
style: match some styling

ajatdarojat45 committed:
Merge pull request #25 from tigadika/feature/setting-tab

Feature/setting tab

ajatdarojat45 committed:
chore(release): 0.0.13

ajatdarojat45 commented:
Hi @SilvinWillemsen,

In this PR we also add:
- Settings tab, fetch settings, and update simulation.
- Add monaco editor to edit setting with json.

SilvinWillemsen commented:
Hi @ajatdarojat45, @tigadika & @fachrihawari ,

Thanks for the changes! I must say that the interaction with the sources/receivers works very smoothly :D Especially selecting sources/receivers in the 3D viewport works much more smoothly than the previous version. The settings tab also looks very good! 

A few comments:

**Sources/receivers**
- When a source is too close to a receiver, the error should be “Too close to a receiver” 
- Sources are allowed to be close together; this is no problem. (Same for receivers, but this seems to already have been implemented.
- (Low priority:) The source/receiver can still be selected after a mouse drag event in the viewport. Only a mouse click should deselect them.
- (Low priority:) Could the source/receiver turn red (in the viewport) if there is an error?

**Settings**
- (just a compliment: ) Very nice that you include the ranges in the UI! Makes it very user-friendly :)
- I can fill in a value that is out of the range. (Try to type 0,001 in the Impulse response length setting)
- The heading should be “Settings” not “Sources” :)
- Also, it is not possible for me to empty the textbox and start typing. Could the range check (+ send to backend) be only upon deselection of the textbox / a return key press?
- (Low priority:) Judging by the “Settings saved” prompt, it seems like the setting gets immediately sent to the backend. This is perfectly fine, but maybe we can only show the prompt something like one second after the latest change. 
    - User clicks a few times on an arrow in the textbox.
    - Current behaviour: “Settings saved” prompt gets shown every time
    - Wanted behaviour: “Settings saved” prompt gets shown one second after the last change.
- (Low priority:) The three-dot button can be a button “Open as JSON” or the like.
- (Low priority:) The reset button looks like it is disabled

**Groups**
- (Low priority:) The trash icon for the Ungrouped group should only show if it has projects.

Quick question, what does this mean: "Feat: Added lazy query for simulations by model ID and updated cache invalidation tags."?

fachrihawari commented:
Hi @SilvinWillemsen, sorry to jump in. 

> Quick question, what does this mean: "Feat: Added lazy query for simulations by model ID and updated cache invalidation tags."?

The message means it’s a fix for a bug where the newly created simulation isn’t selected automatically.


Thats a fix for your comment below.
<img width="841" height="94" alt="Screenshot 2025-10-19 at 17 34 06" src="https://github.com/user-attachments/assets/3e22a9a1-e5fe-487a-95a5-1774e4aec7eb" />

fachrihawari committed:
fix: prevent rendering when there are no projects in "NONE" group

ajatdarojat45 committed:
Merge pull request #37 from fachrihawari/fix/trash-ungroupped-always-shown

fix: Hide trash when Ungrouped section is empty

tigadika committed:
fix: different validation type for source and receiver when close together

tigadika committed:
fix: only register click event to (de)select sources/receivers

tigadika committed:
fix: make source&receivers turns red in viewport when not valid

tigadika committed:
fix: settings validation to not exceed the range input and not call update

tigadika committed:
fix: change debounce value for update settings to 1 second instead of 300ms

tigadika committed:
style: change some styling in json editor

ajatdarojat45 committed:
Merge pull request #39 from tigadika/fix/sources-receivers

Fix/sources receivers

ajatdarojat45 committed:
chore(release): 0.0.14

ajatdarojat45 commented:
Hi @SilvinWillemsen,

We’ve updated the PR based on your notes. Here’s a summary of the changes:
- change debounce value for update settings to 1 second instead of 300ms
- different validation type for source and receiver when close together
- make source&receivers turns red in viewport when not valid
- only register click event to (de)select sources/receivers
- prevent rendering when there are no projects in "NONE" group
- settings validation to not exceed the range input and not call update

> somehow, in settings, user could still inputting out of range value, but if user control using input increment & decrement (arrow up and down), it will not exceed the range. For this, I implemented alternative solution: make the input box red and the range info red too if not valid.

cc: @fachrihawari @tigadika

SilvinWillemsen commented:
HI @ajatdarojat45, @fachrihawari & @tigadika,

Thanks for the changes! All the textboxes seem to work fine, except the reflection coefficient. See the attached [video](https://github.com/user-attachments/assets/24fb8368-5f88-4c3c-8823-34219977df2f) where I press backspace a lot. For the last 9 it works, but once I remove the first 9, it snaps back to 0,99.

Furthermore, when I'm editing a textbox, the Settings saved prompt still shows up a lot. Could we only have it show up when I'm actually done editing the text (i.e., only save the setting when I'm done editing, not upon every change)? (Low priority)

Also, I'm not sure how the "outside model bounds" validation currently works, but I tried a different model, and it seems to only check the outer bounds of the model. If there is a gap in the model (see image) it doesn't correctly validate
<img width="349" height="369" alt="Screenshot 2025-10-20 at 07 58 17" src="https://github.com/user-attachments/assets/69bdf901-02ab-4705-bfea-d88a97ddb763" />

I understand that this can be a hard problem to solve, so if it's too time-consuming for now, please leave it as is.

As these are only minor bugs, I'll merge this pull request now. If you could make note of these and fix them at a later point, that would be great :)

Best,
Silvin
